Give an option to not connect offline Discord users

Sazpaimon opened the pull request.

Sazpaimon committed:
Give an option to not connect offline Discord users

jlu5 commented on README.md:
```diff
@@ -41,6 +41,10 @@ You can also install these dependencies via pip (for Python 3) using: `pip3 inst
                 #webhooks_agent: Webhooks!webhooks@discord/webhooks-agent
+                # If disabled, users that are marked as "Invisible" or "Offline" will not be joined to 
+                # linked channels until they come online. Useful if you have many offline users compared 
+                # to online ones. Note that if this is disabled, PMs cannot be sent to offline Discord users
```
```yaml
                    # This defaults to true if not set.
```

jlu5 commented on protocols/discord.py:
```diff
@@ -227,6 +227,10 @@ def _burst_new_client(self, guild, member, pylink_netobj):
             return
 
+        if not pylink_netobj.serverdata.get('join_offline_users', True) and \
+                member.user.presence in [DiscordStatus.OFFLINE, DiscordStatus.INVISIBLE, UNSET]:
```
Could we move this into a helper function like `_should_burst_user()`? I believe it's possible to invert the logic as follows:

```python
return pylink_netobj.serverdata.get('join_offline_users', True) or \
    member.user.presence not in [DiscordStatus.OFFLINE, DiscordStatus.INVISIBLE, UNSET]
```

jlu5 reviewed (changes requested):
I'm a bit concerned about desyncs here:

1) If the `join_offline_users` option value changes while a guild is connected, things could go badly. For the sake of simplicity, we could perhaps store that value when we initialize a subserver and don't allow it to be changed without a reconnect.

2) I believe the check in `_update_channel_presence` should be updated to silently ignore missing users if this option is disabled, instead of raising an error.

Sazpaimon committed:
Store `join_offline_users` as a subserver property so it can't be changed on rehash
Move duplicate code to a helper function
Prevent channel updates from logging errors on offline users

jlu5 commented on README.md:
```diff
@@ -41,6 +41,12 @@ You can also install these dependencies via pip (for Python 3) using: `pip3 inst
+                # If disabled, users that are marked as "Invisible" or "Offline" will not be joined to 
+                # linked channels until they come online. Useful if you have many offline users compared 
+                # to online ones. Note that if this is disabled, PMs cannot be sent to offline Discord users
+                # Changes to this setting take place on restart
```
Mm, more correct would be:

```yaml
    # Changes to this setting require a reconnect to apply.
```

jlu5 commented on protocols/discord.py:
```diff
@@ -482,7 +495,14 @@ def _update_user_status(self, guild_id, uid, presence):
+    def _should_burst_user(pylink_netobj, user):
+        return not pylink_netobj.join_offline_users and \
+               user.presence != UNSET and \
+               user.presence.status not in [DiscordStatus.OFFLINE, DiscordStatus.INVISIBLE]
```
Could we rewrite this as an `or`? I think it's easier to read

```python
    # Burst users only if join_offline_users is enabled or their status is online/away/DND
    return pylink_netobj.join_offline_users or (user.presence != UNSET and \
        user.presence.status not in [DiscordStatus.OFFLINE, DiscordStatus.INVISIBLE])
```

Sazpaimon replied:
I personally prefer if statements being only one level deep, but I can swing either way, here

jlu5 replied:
I think logically clearer is better

jlu5 commented on README.md:
```diff
@@ -41,6 +41,12 @@ You can also install these dependencies via pip (for Python 3) using: `pip3 inst
+                # linked channels until they come online. Useful if you have many offline users compared 
+                # to online ones. Note that if this is disabled, PMs cannot be sent to offline Discord users
+                # Changes to this setting take place on restart
+                # This defaults to true if not set.
```
And we can put this with the line above
